src/protocol.rs

```diff
@@ -1623,9 +1641,29 @@ impl CryptoServer {
             // ICR7
             peer.session()
                 .insert(self, core.enter_live(self, HandshakeRole::Responder)?)?;
+            // TODO: This should be part of the protocol specification.
+            peer.hs().take(self);
```

Comment:
I do not understand how this is related to the race condition.

Reply:
This is what actually fixes the race condition; it aborts any ongoing handshake from initiator perspective.